File: java_tests/src/main/java/org/infinispan/client/hotrod/configuration/Configuration.java
package org.infinispan.client.hotrod.configuration;

import java.lang.ref.WeakReference;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.infinispan.client.hotrod.impl.consistenthash.ConsistentHash;
import org.infinispan.client.hotrod.impl.transport.TransportFactory;
import org.infinispan.client.hotrod.impl.transport.tcp.RequestBalancingStrategy;
import org.infinispan.client.hotrod.configuration.ConnectionPoolConfiguration;
import org.infinispan.client.hotrod.configuration.SslConfiguration;
import org.infinispan.commons.configuration.BuiltBy;
import org.infinispan.commons.marshall.Marshaller;

/**
 * Configuration.
 *
 * @author Tristan Tarrant
 * @since 5.3
 */
@BuiltBy(ConfigurationBuilder.class)
public class Configuration {

   private final ExecutorFactoryConfiguration asyncExecutorFactory;
   private final Class<? extends RequestBalancingStrategy> balancingStrategy;
   private final WeakReference<ClassLoader> classLoader;
   private final Class<? extends ConsistentHash>[] consistentHashImpl;
   private final Class<? extends Marshaller> marshallerClass;
   private final Marshaller marshaller;
   private final List<ServerConfiguration> servers;
   private final Class<? extends TransportFactory> transportFactory;
   
   private cli.Infinispan.HotRod.Config.Configuration jniConfiguration;

   public Configuration(cli.Infinispan.HotRod.Config.Configuration jniConfiguration) {
      super();
      this.jniConfiguration = jniConfiguration;
      this.asyncExecutorFactory = null;
      this.balancingStrategy = null;
      this.classLoader = null;
      this.consistentHashImpl = null;
      this.marshallerClass = null;
      this.marshaller = null;
      this.servers = null;
      this.transportFactory = null;
   }

   Configuration(ExecutorFactoryConfiguration asyncExecutorFactory, Class<? extends RequestBalancingStrategy> balancingStrategy, ClassLoader classLoader,
         ConnectionPoolConfiguration connectionPool, int connectionTimeout, Class<? extends ConsistentHash>[] consistentHashImpl, boolean forceReturnValues, int keySizeEstimate, Class<? extends Marshaller> marshallerClass,
         boolean pingOnStartup, String protocolVersion, List<ServerConfiguration> servers, int socketTimeout, SslConfiguration ssl, boolean tcpNoDelay,
         Class<? extends TransportFactory> transportFactory, int valueSizeEstimate) {
       this.jniConfiguration = null; // new cli.Infinispan.HotRod.Config.ConfigurationBuilder().Build();
      this.asyncExecutorFactory = asyncExecutorFactory;
      this.balancingStrategy = balancingStrategy;
      this.classLoader = new WeakReference<ClassLoader>(classLoader);
      this.consistentHashImpl = consistentHashImpl;
      this.marshallerClass = marshallerClass;
      this.marshaller = null;
      this.servers = Collections.unmodifiableList(servers);
      this.transportFactory = transportFactory;
   }

   Configuration(ExecutorFactoryConfiguration asyncExecutorFactory, Class<? extends RequestBalancingStrategy> balancingStrategy, ClassLoader classLoader,
         ConnectionPoolConfiguration connectionPool, int connectionTimeout, Class<? extends ConsistentHash>[] consistentHashImpl, boolean forceReturnValues, int keySizeEstimate, Marshaller marshaller,
         boolean pingOnStartup, String protocolVersion, List<ServerConfiguration> servers, int socketTimeout, SslConfiguration ssl, boolean tcpNoDelay,
         Class<? extends TransportFactory> transportFactory, int valueSizeEstimate) {
       this.jniConfiguration = null; // new cli.Infinispan.HotRod.Config.ConfigurationBuilder().Build();
      this.asyncExecutorFactory = asyncExecutorFactory;
      this.balancingStrategy = balancingStrategy;
      this.classLoader = new WeakReference<ClassLoader>(classLoader);
      this.consistentHashImpl = consistentHashImpl;
      this.marshallerClass = null;
      this.marshaller = marshaller;
      this.servers = Collections.unmodifiableList(servers);
      this.transportFactory = transportFactory;
   }

   public cli.Infinispan.HotRod.Config.Configuration getJniConfiguration() {
      return jniConfiguration;
   }

   public ExecutorFactoryConfiguration asyncExecutorFactory() {
      return asyncExecutorFactory;
   }

   public Class<? extends RequestBalancingStrategy> balancingStrategy() {
      return balancingStrategy;
   }

   public ClassLoader classLoader() {
      return classLoader.get();
   }

   public ConnectionPoolConfiguration connectionPool() {
       throw new UnsupportedOperationException();
      // return new ConnectionPoolConfiguration(this.jniConfiguration.ConnectionPool());
   }

   public int connectionTimeout() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.ConnectionTimeout();
   }

   public Class<? extends ConsistentHash>[] consistentHashImpl() {
      return consistentHashImpl;
   }

   public Class<? extends ConsistentHash> consistentHashImpl(int version) {
      return consistentHashImpl[version-1];
   }

   public boolean forceReturnValues() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.ForceReturnValues();
   }

   public int keySizeEstimate() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.KeySizeEstimate();
   }

   public Marshaller marshaller() {
      return marshaller;
   }

   public Class<? extends Marshaller> marshallerClass() {
      return marshallerClass;
   }

   public boolean pingOnStartup() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.PingOnStartup();
   }

   public String protocolVersion() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.ProtocolVersion();
   }

   public List<ServerConfiguration> servers() {
      return servers;
   }

   public int socketTimeout() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.SocketTimeout();
   }

   public SslConfiguration ssl() {
       throw new UnsupportedOperationException();
      // return new SslConfiguration(this.jniConfiguration.Ssl());
   }

   public boolean tcpNoDelay() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.TcpNoDelay();
   }

   public Class<? extends TransportFactory> transportFactory() {
      return transportFactory;
   }

   public int valueSizeEstimate() {
       throw new UnsupportedOperationException();
      // return this.jniConfiguration.ValueSizeEstimate();
   }

   public int maxRetries() {
       throw new UnsupportedOperationException();
      //return this.jniConfiguration.getMaxRetries();
   }

   @Override
   public String toString() {
      return "Configuration [asyncExecutorFactory=" + asyncExecutorFactory + ", balancingStrategy=" + balancingStrategy + ", classLoader=" + classLoader + ", connectionPool="
            + connectionPool() + ", connectionTimeout=" + connectionTimeout() + ", consistentHashImpl=" + Arrays.toString(consistentHashImpl) + ", forceReturnValues="
            + forceReturnValues() + ", keySizeEstimate=" + keySizeEstimate() + ", marshallerClass=" + marshallerClass + ", marshaller=" + marshaller + ", pingOnStartup="
            + pingOnStartup() + ", protocolVersion=" + protocolVersion() + ", servers=" + servers + ", socketTimeout=" + socketTimeout() + ", ssl=" + ssl() + ", tcpNoDelay=" + tcpNoDelay()
            + ", transportFactory=" + transportFactory + ", valueSizeEstimate=" + valueSizeEstimate() + "]";
   }
}

File: src/Infinispan.HotRod/Config/Configuration.cs
using System.Collections;
using System.Collections.Generic;
using Infinispan.HotRod.SWIG;

namespace Infinispan.HotRod.Config
{
    /// <summary>
    ///   Used to hold the configuration parameters.
    /// </summary>
    public class Configuration
    {
        private Infinispan.HotRod.SWIG.Configuration config;
        private IMarshaller marshaller;
        private AuthenticationConfiguration authConfig;
        internal Configuration(Infinispan.HotRod.SWIG.Configuration config, IMarshaller marshaller, AuthenticationConfiguration authConfig)
        {
            this.config = config;
            this.marshaller = marshaller;
            this.authConfig = authConfig;
        }

        internal Infinispan.HotRod.SWIG.Configuration Config()
        {
            return config;
        }

        /// <summary>
        ///   Retrieve the protocol version.
        /// </summary>
        ///
        /// <returns>a protocol version string</returns>
        public string ProtocolVersion()
        {
            return config.getProtocolVersion();
        }

        /// <summary>
        ///   Retrieve the configuration pool configuration.
        /// </summary>
        ///
        /// <returns>an object holding the connection pool configurations</returns>
        public ConnectionPoolConfiguration ConnectionPool()
        {
            return new ConnectionPoolConfiguration(config.ConnectionPool());
        }

        /// <summary>
        ///   Retrives the connection timeout.
        /// </summary>
        public int ConnectionTimeout()
        {
            return config.getConnectionTimeout();
        }

        /// <summary>
        ///   Retrives the state of the force return values flag.
        /// </summary>
        public bool ForceReturnValues()
        {
            return config.isForceReturnValue();
        }

        /// <summary>
        ///   Retrieves the configured estimated size for keys.
        /// </summary>
        public int KeySizeEstimate()
        {
            return config.getKeySizeEstimate();
        }

        /// <summary>
        ///   Retrieves a list of objects holding the server configurations.
        /// </summary>
        public IList<ServerConfiguration> Servers()
        {
            IList<Infinispan.HotRod.SWIG.ServerConfiguration> res = new List<Infinispan.HotRod.SWIG.ServerConfiguration>();
            config.GetServersMapConfiguration().TryGetValue("DEFAULT_CLUSTER_NAME",out res);
            IList<ServerConfiguration> ret = new List<ServerConfiguration>();
            foreach(Infinispan.HotRod.SWIG.ServerConfiguration s in res) {
                ret.Add(new ServerConfiguration(s));
            }
            return ret;
        }


        /// <summary>
        ///   Retrieves the value configured for socket timeouts.
        /// </summary>
        public int SocketTimeout()
        {
            return config.getSocketTimeout();
        }

        /// <summary>
        ///   Retrieves an object holding the SSL specific configurations.
        /// </summary>
        public SslConfiguration Ssl()
        {
            return new SslConfiguration(config.Ssl());
        }

        /// <summary>
        ///   Retrieves the state of the TCP no delay flag.
        /// </summary>
        public bool TcpNoDelay()
        {
            return config.isTcpNoDelay();
        }

        /// <summary>
        ///   Retrieves the configured estimates size for values.
        /// </summary>
        public int ValueSizeEstimate()
        {
            return config.getValueSizeEstimate();
        }

        /// <summary>
        ///   Retrieves the configured max retries value
        /// </summary>
        public int MaxRetries()
        {
            return config.getMaxRetries();
        }

        /// <summary>
        ///   Retrieves the the configured marshaller.
        /// </summary>
        public IMarshaller Marshaller()
        {
            return marshaller;
        }
    }
}